### PR TITLE
feat(assets): scope Data Consumer asset visibility to DP-linked assets (#347)

### LINKED ISSUE #347: [Feature]: Scope Data Consumer asset visibility to assets linked from accessible Data Products
## Context

A customer using Ontos with role-based access flagged a visibility gap in how Assets are exposed to Data Consumers.

## Current behavior

Two failure modes depending on whether the Data Consumer role has the `assets` feature granted:

1. **`assets` feature granted (READ_ONLY):** Data Consumers see *all* assets in the Asset Explorer, regardless of whether those assets are linked to a Data Product they can access. Over-shares.
2. **`assets` feature not granted:** Data Consumers cannot see linked assets *inside* a Data Product detail view either — the `Linked Assets` panel falls back to empty because the asset fetch is gated by the same feature flag. Under-shares.

Neither mode matches intent.

## Desired behavior

A Data Consumer should see assets *scoped to the Data Products they can access*:

- In the **DP detail view's `Linked Assets` panel** — show assets linked to *this* DP, regardless of `assets` feature state.
- In the **Asset Explorer** (if exposed to consumers at all) — show only assets linked to at least one DP the consumer can access.

## Why this matters

- Without the feature flag, consumers can't browse what they're entitled to.
- With the feature flag, consumers see assets they have no business seeing — a least-privilege violation in deployments that take role boundaries seriously.

## Implementation considerations (not prescriptive)

Two candidate paths:

- **Frontend filter** — smaller change, but leaks data over the wire and through direct API calls.
- **Backend scoped endpoint / role-aware filter** — add DP-scoping to `GET /api/assets` (or introduce a `GET /api/data-products/{id}/assets` endpoint). The consumer's accessible-DP set drives the asset query. Larger change, secure by default.

Backend route is the safer of the two; worth a quick scoping discussion before implementation.

## Out of scope

- Ontology changes
- Permission-model changes for non-Consumer roles

## Acceptance criteria

- [ ] Data Consumers see `Linked Assets` in DP detail view even without `assets` feature granted
- [ ] Data Consumers do not see assets unrelated to DPs they can access, regardless of `assets` feature state
- [ ] Existing behavior for Data Producers / Admins unchanged
- [ ] Tests cover both the linked-from-DP path and the broader Asset Explorer path

### PR DESCRIPTION
**Pull request by mvkonchits-db**
Closes #347.

## Summary

Scopes Data Consumer asset visibility to assets linked to Data Products they can access. Producers + Admins (`assets:READ_WRITE`+) keep full visibility — existing behaviour preserved. Adds a dedicated `GET /api/data-products/{id}/assets` endpoint so the DP detail view's Linked Assets panel works without the `assets` feature granted to Consumers.

## Design

Backend route, not frontend filter — issue #347 explicitly recommended the secure-by-default path.

The discriminator is `assets:READ_WRITE+`:
- **READ_WRITE+** (Producers, Admins) → unscoped, full asset list
- **Below READ_WRITE** (Consumers, users without `assets` feature) → scoped via `entity_relationships` to assets linked to a DataProduct or OutputPort the user can list

Scope resolution reuses `data_products_manager.list_products(is_admin=False)` so there is one source of truth for "which DPs can this user see."

## Files changed

| File | Purpose |
|---|---|
| `src/backend/src/common/data_product_asset_scope.py` (new) | DP→asset resolution helpers |
| `src/backend/src/repositories/assets_repository.py` | `restrict_to_ids` filter on the asset query |
| `src/backend/src/controller/assets_manager.py` | Plumb `restrict_to_ids` through |
| `src/backend/src/routes/assets_routes.py` | In-handler authorization (replaces blanket `PermissionChecker(assets, READ_ONLY)`) on `GET /api/assets` and `GET /api/assets/{id}` |
| `src/backend/src/routes/data_product_routes.py` | New `GET /api/data-products/{id}/assets`, gated by `data-products:READ_ONLY` (not `assets`) |
| `src/backend/src/tests/unit/test_data_product_asset_scope.py` (new) | 18 unit tests |

## Acceptance criteria (from #347)

- [x] Data Consumers see Linked Assets in DP detail view even without `assets` feature granted
- [x] Data Consumers do not see assets unrelated to DPs they can access, regardless of `assets` feature state
- [x] Existing behavior for Data Producers / Admins unchanged
- [x] Tests cover both the linked-from-DP path and the broader Asset Explorer path

### Unit tests
18/18 passing — admin/non-admin paths, empty DP set, multi-DP dedup, port-mediated linkage, `restrict_to_ids` empty/None edge cases.

### E2E on a real workspace deployment
Three deploys against a demo workspace to converge:

| Scenario | Endpoint | Expected | Actual |
|---|---|---|---|
| Admin (regression) | `GET /api/assets` | full set incl. orphan | ✅ 110 items |
| Admin (regression) | `GET /api/assets/{orphan_id}` | 200 | ✅ 200 |
| Consumer w/ `assets:None` (AC1) | `GET /api/data-products/{id}/assets` | 200 with linked items | ✅ 200, 2 items |
| Consumer (AC2) | `GET /api/assets` | scoped, orphan invisible | ✅ 14 items |
| Consumer (AC2) | `GET /api/assets/{orphan_id}` | 403 | ✅ 403 |
| Consumer (AC2) | `GET /api/assets/{dp_linked_id}` | 200 | ✅ 200 |

Test orphan asset created and deleted as part of the run.

## Non-obvious findings (preserved for reviewers)

- `is_user_admin(groups, settings)` was the natural discriminator on paper, but groups arrive as `null` from request headers in some deployments — that path returns False for actual admins. Switched to `authorization_manager.has_permission(perms, "assets", READ_WRITE)` which respects per-user role overrides + email-as-implicit-group. Caught by the E2E run, not by unit tests.
- App-state key is `authorization_manager`, not `auth_manager`. Caught by the same E2E.
- Fail-closed when `DataProductsManager` is unavailable (early-startup or test scaffolding) — scoped users get an empty list, not a 500.

## Out of scope

- Ontology changes
- Permission-model changes for non-Consumer roles
- Frontend changes (the existing DP detail Linked Assets panel already calls a per-DP endpoint shape; this PR makes that endpoint exist with the right authorization)

### CODE REVIEW

File: src/backend/src/common/data_product_asset_scope.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
+"""Helpers for scoping asset visibility to Data Products a user can access.
```
Comment (larsgeorge-db) [inline src/backend/src/common/data_product_asset_scope.py]: Why is this a helper and why is it accessing the database directly? Could this not be a method in the proper controller?

Comment (mvkonchits-db) [reply src/backend/src/common/data_product_asset_scope.py]: Refactored — DB queries moved to `DataProductRepository.get_output_port_ids_for_products` and `EntityRelationshipRepository.get_asset_ids_linked_to_products`; orchestration (`resolve_accessible_asset_ids`, `is_asset_accessible`) moved to `AssetsManager`; `DataProductsManager` exposes `list_linked_asset_ids_for_products`. Routes call manager methods. `src/backend/src/common/data_product_asset_scope.py` deleted. No behavioural change — 16 unit tests still passing in `test_data_product_asset_scoping.py`, full unit suite green (888 passed). Commit: 8ff9bda96e1e8f1f88b68e3b71504ff595201030.

**Review by larsgeorge-db — APPROVED** [review]
